src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl_ModelComputation.cs

```diff
@@ -0,0 +1,257 @@
+
+            if (ErrorHandler.Failed(CodeWindow.GetLastActiveView(out var textView)))
+                return null;
+
+            return EditorAdaptersFactoryService.GetWpfTextView(textView);
+        }
+
+        /// <summary>
+        /// Starts a new task to compute the UI model.
+        /// </summary>
+        private void StartComputeUIModelTask()
+        {
+            // 'true' value is unused.  this just signals to the queue that we have work to do.
+            _computeUIModelQueue.AddWork(true);
+        }
+
+        /// <summary>
+        /// Creates the UI model.
+        /// </summary>
+        private async ValueTask<DocumentSymbolModel?> ComputeUIModelAsync(ImmutableSegmentedList<bool> unused, CancellationToken cancellationToken)
+        {
+            // Jump to the UI thread to get the currently active text view.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return null;
+
+            var originalSnapshot = activeTextView.TextSnapshot;
+            var textBuffer = activeTextView.TextBuffer;
+
+            var filePath = GetFilePath();
+            if (filePath is null)
+                return null;
+
+            // Ensure we switch to the threadpool before calling ComputeUIModelAsync. It ensures
+            // that fetching and processing the document model is not done on the UI thread.
+            await TaskScheduler.Default;
+
+            var model = await ComputeUIModelAsync().ConfigureAwait(false);
+
+            if (model is not null)
+                StartUpdateUIModelTask();
+
+            return model;
+
+            async Task<DocumentSymbolModel?> ComputeUIModelAsync()
+            {
+                var response = await DocumentOutlineHelper.DocumentSymbolsRequestAsync(
+                    textBuffer, LanguageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+
+                if (response is null)
+                    return null;
+
+                var responseBody = response.ToObject<DocumentSymbol[]>();
+                if (responseBody is null)
+                    return null;
+
+                var documentSymbols = DocumentOutlineHelper.GetNestedDocumentSymbols(responseBody);
+                var documentSymbolItems = DocumentOutlineHelper.GetDocumentSymbolItems(documentSymbols);
+                return new DocumentSymbolModel(documentSymbolItems, originalSnapshot);
+            }
+
+            string? GetFilePath()
+            {
+                ThreadingContext.ThrowIfNotOnUIThread();
+                if (EditorAdaptersFactoryService.GetBufferAdapter(textBuffer) is IPersistFileFormat persistFileFormat &&
+                    ErrorHandler.Succeeded(persistFileFormat.GetCurFile(out var filePath, out _)))
+                {
+                    return filePath;
+                }
```

Comment:
interesting!  @jasonmalinowski is this the appropriate way to do this?  Should we not be going from ITextSnapshot (or ITextBuffer) to workspace Document instance and using that as the source of truth?

Reply:
I'd normally say yes, we should be doing that. In this case though, since I'd expect this code to eventually migrate to the VS LSP client, there's no reason to tie to the workspace model here.

That said, it might be easier/saner just to go the ITextDocument and get the path from there, which might match better what the LSP client is doing. But if this matches current LSP client behavior, then I won't complain.